// @hocuspocus/extension-webhook/dist/hocuspocus-webhook.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let node_crypto = require("node:crypto");
let _hocuspocus_common = require("@hocuspocus/common");
let _hocuspocus_transformer = require("@hocuspocus/transformer");

//#region packages/extension-webhook/src/index.ts
let Events = /* @__PURE__ */ function(Events) {
	Events["onChange"] = "change";
	Events["onConnect"] = "connect";
	Events["onCreate"] = "create";
	Events["onDisconnect"] = "disconnect";
	return Events;
}({});
var Webhook = class {
	/**
	* Constructor
	*/
	constructor(configuration) {
		this.configuration = {
			debounce: 2e3,
			debounceMaxWait: 1e4,
			secret: "",
			transformer: _hocuspocus_transformer.TiptapTransformer,
			url: "",
			events: [Events.onChange]
		};
		this.debounced = /* @__PURE__ */ new Map();
		this.configuration = {
			...this.configuration,
			...configuration
		};
		if (!this.configuration.url) throw new Error("url is required!");
	}
	/**
	* Create a signature for the response body
	*/
	createSignature(body) {
		return `sha256=${(0, node_crypto.createHmac)("sha256", this.configuration.secret).update(body).digest("hex")}`;
	}
	/**
	* debounce the given function, using the given identifier
	*/
	debounce(id, func) {
		const old = this.debounced.get(id);
		const start = old?.start || Date.now();
		const run = () => {
			this.debounced.delete(id);
			func();
		};
		if (old?.timeout) clearTimeout(old.timeout);
		if (Date.now() - start >= this.configuration.debounceMaxWait) return run();
		this.debounced.set(id, {
			start,
			timeout: setTimeout(run, this.configuration.debounce)
		});
	}
	/**
	* Send a request to the given url containing the given data
	*/
	async sendRequest(event, payload) {
		const json = JSON.stringify({
			event,
			payload
		});
		const response = await fetch(this.configuration.url, {
			method: "POST",
			body: json,
			headers: {
				"X-Hocuspocus-Signature-256": this.createSignature(json),
				"Content-Type": "application/json"
			}
		});
		if (!response.ok) throw new Error(`Webhook request to ${this.configuration.url} failed with status ${response.status}`);
		const text = await response.text();
		const data = (response.headers.get("content-type") ?? "").includes("application/json") && text ? JSON.parse(text) : text;
		return {
			status: response.status,
			data
		};
	}
	/**
	* onChange hook
	*/
	async onChange(data) {
		if (!this.configuration.events.includes(Events.onChange)) return;
		const save = async () => {
			try {
				await this.sendRequest(Events.onChange, {
					document: this.configuration.transformer.fromYdoc(data.document),
					documentName: data.documentName,
					context: data.context,
					requestHeaders: data.requestHeaders,
					requestParameters: Object.fromEntries(data.requestParameters.entries())
				});
			} catch (e) {
				console.error(`Caught error in extension-webhook: ${e}`);
			}
		};
		if (!this.configuration.debounce) return save();
		this.debounce(data.documentName, save);
	}
	/**
	* onLoadDocument hook
	*/
	async onLoadDocument(data) {
		if (!this.configuration.events.includes(Events.onCreate)) return;
		try {
			const document = (await this.sendRequest(Events.onCreate, {
				documentName: data.documentName,
				requestHeaders: data.requestHeaders,
				requestParameters: Object.fromEntries(data.requestParameters.entries())
			})).data;
			if (!document) return;
			for (const fieldName in document) if (data.document.isEmpty(fieldName)) data.document.merge(this.configuration.transformer.toYdoc(document[fieldName], fieldName));
		} catch (e) {
			console.error(`Caught error in extension-webhook: ${e}`);
		}
	}
	/**
	* onConnect hook
	*/
	async onConnect(data) {
		if (!this.configuration.events.includes(Events.onConnect)) return;
		try {
			return (await this.sendRequest(Events.onConnect, {
				documentName: data.documentName,
				requestHeaders: data.requestHeaders,
				requestParameters: Object.fromEntries(data.requestParameters.entries())
			})).data;
		} catch (e) {
			console.error(`Caught error in extension-webhook: ${e}`);
			throw _hocuspocus_common.Forbidden;
		}
	}
	async onDisconnect(data) {
		if (!this.configuration.events.includes(Events.onDisconnect)) return;
		try {
			await this.sendRequest(Events.onDisconnect, {
				documentName: data.documentName,
				requestHeaders: data.requestHeaders,
				requestParameters: Object.fromEntries(data.requestParameters.entries()),
				context: data.context
			});
		} catch (e) {
			console.error(`Caught error in extension-webhook: ${e}`);
		}
	}
};

//#endregion
exports.Events = Events;
exports.Webhook = Webhook;
